File: frontend/src/engine/capabilities/crossTabulate.ts
/**
 * crossTabulate – Build a cross-tab (contingency) matrix from raw data.
 *
 * Example: Count candidates per role per department
 *   Row field:    Department
 *   Column field: Role
 *   Value:        Candidate name (count)
 *
 * Output is a static matrix written to outputRange.
 */

import { CapabilityMeta, CrossTabulateParams, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";
import { resolveRange } from "./rangeUtils";

const meta: CapabilityMeta = {
  action: "crossTabulate",
  description: "Build a cross-tab matrix counting or summing values across two dimensions",
  mutates: true,
  affectsFormatting: false,
};

async function handler(
  context: Excel.RequestContext,
  params: CrossTabulateParams,
  options: ExecutionOptions,
): Promise<StepResult> {
  const { sourceRange, rowField, columnField, valueField, aggregation = "count", outputRange } = params;

  if (options.dryRun) {
    return { stepId: "", status: "success", message: `Would cross-tabulate ${sourceRange} (row: col ${rowField}, col: col ${columnField})` };
  }

  options.onProgress?.("Reading source data...");
  const srcRng = resolveRange(context, sourceRange);
  srcRng.load("values");
  await context.sync();

  const data = (srcRng.values ?? []) as (string | number | boolean | null)[][];
  if (data.length < 2) return { stepId: "", status: "success", message: "Not enough data." };

  const rowIdx = rowField - 1;
  const colIdx = columnField - 1;
  const valIdx = valueField - 1;

  options.onProgress?.("Building cross-tab matrix...");

  // Collect unique row keys and column keys (in order of first appearance)
  const rowKeys: string[] = [];
  const colKeys: string[] = [];
  const rowSet = new Set<string>();
  const colSet = new Set<string>();

  for (let r = 1; r < data.length; r++) {
    const rk = String(data[r][rowIdx] ?? "");
    const ck = String(data[r][colIdx] ?? "");
    if (!rowSet.has(rk)) { rowSet.add(rk); rowKeys.push(rk); }
    if (!colSet.has(ck)) { colSet.add(ck); colKeys.push(ck); }
  }

  // Accumulate values: matrix[rowKey][colKey] = { sum, count }
  type Cell = { sum: number; count: number };
  const matrix: Map<string, Map<string, Cell>> = new Map();
  for (const rk of rowKeys) {
    const row = new Map<string, Cell>();
    for (const ck of colKeys) row.set(ck, { sum: 0, count: 0 });
    matrix.set(rk, row);
  }

  for (let r = 1; r < data.length; r++) {
    const rk = String(data[r][rowIdx] ?? "");
    const ck = String(data[r][colIdx] ?? "");
    const v = Number(data[r][valIdx]) || (aggregation === "count" ? 1 : 0);
    const cell = matrix.get(rk)?.get(ck);
    if (cell) { cell.sum += v; cell.count++; }
  }

  // Build output 2D array
  const outRows: (string | number)[][] = [];
  // Header row: blank top-left, then column keys
  outRows.push(["", ...colKeys, "Total"]);
  // Data rows
  for (const rk of rowKeys) {
    const row: (string | number)[] = [rk];
    let rowTotal = 0;
    for (const ck of colKeys) {
      const cell = matrix.get(rk)?.get(ck) ?? { sum: 0, count: 0 };
      const v = aggregation === "count" ? cell.count : aggregation === "sum" ? cell.sum : cell.count ? cell.sum / cell.count : 0;
      row.push(v);
      rowTotal += aggregation === "count" ? cell.count : cell.sum;
    }
    row.push(rowTotal);
    outRows.push(row);
  }
  // Total row
  const totalsRow: (string | number)[] = ["Total"];
  let grandTotal = 0;
  for (let ci = 0; ci < colKeys.length; ci++) {
    let colTotal = 0;
    for (const rk of rowKeys) {
      const cell = matrix.get(rk)?.get(colKeys[ci]) ?? { sum: 0, count: 0 };
      colTotal += aggregation === "count" ? cell.count : cell.sum;
    }
    totalsRow.push(colTotal);
    grandTotal += colTotal;
  }
  totalsRow.push(grandTotal);
  outRows.push(totalsRow);

  const outRng = resolveRange(context, outputRange);
  outRng.getResizedRange(outRows.length - 1, outRows[0].length - 1).values = outRows as any;
  await context.sync();

  return {
    stepId: "",
    status: "success",
    message: `Cross-tab: ${rowKeys.length} rows × ${colKeys.length} columns (${aggregation}) written to ${outputRange}`,
    outputs: { outputRange },
  };
}

registry.register(meta, handler as any);
export { meta };

File: backend/app/execution/capabilities/cross_tabulate.py
"""crossTabulate — build a cross-tab (contingency) matrix from raw data.

Takes a 1-based row field, column field and value field from the source
range and produces a matrix: unique row keys × unique column keys, with
per-cell aggregation (count / sum / average). Adds a trailing Total column
and Total row. Output is a static 2D block written to the output range.
"""

from __future__ import annotations

from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry
from app.execution.range_utils import resolve_range


def _to_float(x: Any) -> float:
    try:
        if x is None or x == "":
            return 0.0
        return float(x)
    except (TypeError, ValueError):
        return 0.0


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    source = params.get("sourceRange")
    row_field = params.get("rowField")
    column_field = params.get("columnField")
    value_field = params.get("valueField")
    aggregation = (params.get("aggregation") or "count").lower()
    output = params.get("outputRange")

    if not source or not output:
        return {
            "status": "error",
            "message": "crossTabulate requires 'sourceRange' and 'outputRange'.",
        }
    if row_field is None or column_field is None or value_field is None:
        return {
            "status": "error",
            "message": "crossTabulate requires 'rowField', 'columnField', and 'valueField'.",
        }

    if ctx.dry_run:
        return {
            "status": "preview",
            "message": (
                f"Would cross-tabulate {source} (row: col {row_field}, "
                f"col: col {column_field})."
            ),
        }

    try:
        src_rng = resolve_range(ctx.workbook_handle, source)
        raw = src_rng.value
        if raw is None:
            return {"status": "success", "message": "Not enough data.", "outputs": {}}
        if not isinstance(raw, list):
            data: list[list[Any]] = [[raw]]
        elif raw and not isinstance(raw[0], list):
            rows_, _ = src_rng.shape
            data = [list(raw)] if rows_ == 1 else [[v] for v in raw]
        else:
            data = [list(r) for r in raw]

        if len(data) < 2:
            return {"status": "success", "message": "Not enough data.", "outputs": {}}

        row_idx = int(row_field) - 1
        col_idx = int(column_field) - 1
        val_idx = int(value_field) - 1

        # Collect unique row/column keys (preserving first-seen order).
        row_keys: list[str] = []
        col_keys: list[str] = []
        row_set: set[str] = set()
        col_set: set[str] = set()
        for r in range(1, len(data)):
            row = data[r]
            rk = str(row[row_idx]) if row_idx < len(row) and row[row_idx] is not None else ""
            ck = str(row[col_idx]) if col_idx < len(row) and row[col_idx] is not None else ""
            if rk not in row_set:
                row_set.add(rk)
                row_keys.append(rk)
            if ck not in col_set:
                col_set.add(ck)
                col_keys.append(ck)

        # Initialize matrix[rk][ck] = {sum, count}.
        matrix: dict[str, dict[str, dict[str, float]]] = {
            rk: {ck: {"sum": 0.0, "count": 0} for ck in col_keys} for rk in row_keys
        }

        for r in range(1, len(data)):
            row = data[r]
            rk = str(row[row_idx]) if row_idx < len(row) and row[row_idx] is not None else ""
            ck = str(row[col_idx]) if col_idx < len(row) and row[col_idx] is not None else ""
            raw_val = row[val_idx] if val_idx < len(row) else None
            v = _to_float(raw_val)
            if v == 0.0 and aggregation == "count":
                v = 1.0  # count: each row contributes 1
            cell = matrix.get(rk, {}).get(ck)
            if cell is not None:
                cell["sum"] += v
                cell["count"] += 1

        # Build output.
        out_rows: list[list[Any]] = []
        out_rows.append(["", *col_keys, "Total"])

        for rk in row_keys:
            row_out: list[Any] = [rk]
            row_total = 0.0
            for ck in col_keys:
                cell = matrix[rk][ck]
                if aggregation == "count":
                    v = cell["count"]
                elif aggregation == "sum":
                    v = cell["sum"]
                else:  # average
                    v = (cell["sum"] / cell["count"]) if cell["count"] else 0.0
                row_out.append(v)
                row_total += cell["count"] if aggregation == "count" else cell["sum"]
            row_out.append(row_total)
            out_rows.append(row_out)

        totals_row: list[Any] = ["Total"]
        grand_total = 0.0
        for ck in col_keys:
            col_total = 0.0
            for rk in row_keys:
                cell = matrix[rk][ck]
                col_total += cell["count"] if aggregation == "count" else cell["sum"]
            totals_row.append(col_total)
            grand_total += col_total
        totals_row.append(grand_total)
        out_rows.append(totals_row)

        rows = len(out_rows)
        cols = len(out_rows[0])
        out_rng = resolve_range(ctx.workbook_handle, output)
        out_rng.resize(rows, cols).value = out_rows
    except Exception as exc:  # noqa: BLE001
        return {"status": "error", "message": f"crossTabulate failed: {exc}", "error": str(exc)}

    return {
        "status": "success",
        "message": (
            f"Cross-tab: {len(row_keys)} rows × {len(col_keys)} columns "
            f"({aggregation}) written to {output}."
        ),
        "outputs": {"outputRange": output},
    }


registry.register("crossTabulate", handler, mutates=True)
